Add kernel-64k kernelType

jbtrystram opened the pull request:
Since RHEL 9 reverted to 4K memory pages for aarch64, add a way to
switch to a hugepage kernel.
The MachineConfig should contain the following to trigger the kernel
switch:
spec:
  kernelType: 64k-pages

This is exclusive with the `realtime` kernel option.

xref https://issues.redhat.com/browse/COS-2402
This requires https://github.com/openshift/os/pull/1351

travier commented on docs/MachineConfiguration.md:
```diff
@@ -201,7 +201,11 @@ spec:
+In OCP 4.6 the supported extensions is `usbguard`. In OCP 4.8 the supported extensions are `usbguard` and `sandboxed-containers`.  
+In OCP 4.11 the supported extensions are `usbguard`, `sandboxed-containers`, and `kerberos`. 
+In OCP 4.14 the supported extensions are `usbguard`, `sandboxed-containers`, `kerberos`, `ipsec` and `wasm`.
+In OCP 4.15 the supported extensions are `usbguard`, `sandboxed-containers`, `kerberos`, `ipsec`, `wasm` and `kernel-64k`.
```
Let's make a table?

travier commented on pkg/daemon/update.go:
```diff
@@ -1084,6 +1084,7 @@ func getSupportedExtensions() map[string][]string {
 		"kerberos":             {"krb5-workstation", "libkadm5"},
 		"kernel-devel":         {"kernel-devel", "kernel-headers"},
 		"sandboxed-containers": {"kata-containers"},
+		"kernel-64k":           {"kernel-64k"},
```
This does not look like the place it should be added. This is replacing the kernel so it should use the same logic as the kernel-rt extension.

travier commented on test/e2e-single-node/sno_mcd_test.go:
```diff
@@ -196,7 +196,7 @@ func TestExtensions(t *testing.T) {
 				Raw: helpers.MarshalOrDie(ctrlcommon.NewIgnConfig()),
 			},
-			Extensions: []string{"wasm", "ipsec", "usbguard", "kernel-devel", "kerberos"},
+			Extensions: []string{"wasm", "ipsec", "usbguard", "kernel-devel", "kerberos", "kernel-64k"},
```
idem as above

travier commented:
Not an MCO expert but this extension is more a "kernel-rt" extension than a classic extension so it should likely follow a similar code path.

travier commented:
Found:
- https://github.com/openshift/machine-config-operator/blob/3c372bd6fb5da91784fbc582279438e188c5ec83/pkg/daemon/update.go#L1124
- https://docs.openshift.com/container-platform/4.13/post_installation_configuration/machine-configuration-tasks.html#nodes-nodes-rtkernel-arguments_post-install-machine-configuration-tasks

jbtrystram commented:
Updated the PR with a first round of changes, following the `kernel-rt` code path. 

Should I duplicate the e2e test `[TestKernelType](https://github.com/openshift/machine-config-operator/blob/3c372bd6fb5da91784fbc582279438e188c5ec83/test/e2e/mcd_test.go#L195)` and test for the 64k kernel ? Or is this test enough to validate that switching kernels works without testing both cases ? 

Also, please wait a bit while I go through a deployment of this to test on a real cluster :)

travier commented on docs/MachineConfiguration.md:
```diff
@@ -201,7 +201,14 @@ spec:
+| 4.8           |  `usbguard`, `sandboxed-containers`    |
+| 4.11          |  `usbguard`, `sandboxed-containers`, `kerberos`   |
+| 4.14          |  `usbguard`, `sandboxed-containers`, `kerberos`, `ipsec`, `wasm`   |
+| 4.15          |  `usbguard`, `sandboxed-containers`, `kerberos`, `ipsec`, `wasm`, `kernel-64k`   |
```
Thanks for the table conversion :).

I've now realized that the rt extension is not listed there, so this one likely should not be there but in the section above with the rt kernel.

jbtrystram replied:
good catch ! I updated the `### KernelType` section above

cgwalters commented on pkg/daemon/update.go:
```diff
@@ -1143,6 +1145,8 @@ func (dn *CoreOSDaemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig)
 	// of other dependencies we don't want for historical reasons.
 	// kernel-rt also has a split off kernel-rt-kvm subpackage because it's in a separate subscription in RHEL.
 	realtimeKernel := []string{"kernel-rt-core", "kernel-rt-modules", "kernel-rt-modules-extra", "kernel-rt-kvm"}
+	// kernel-64k also has a split off kernel-rt-kvm subpackage because it's in a separate subscription in RHEL.
```
This comment should just be dropped, it doesn't appear to be correct.

cgwalters commented:
/ok-to-test
/approve

jbtrystram commented:
/retest

jlebon commented on docs/MachineConfiguration.md:
```diff
@@ -180,7 +180,7 @@ When a machine boots with `nosmt` Kernel Argument, it disables multi-threading o
 
 This feature is available with OCP 4.4 and onward releases as both `day 1` and `day 2` operation. It allows to choose between traditional and Real Time (RT) kernel on an RHCOS node. Supported values are
-`""` or `default` for traditional kernel and `realtime` for RT kernel.
+`""` or `default` for traditional kernel, `realtime` for RT kernel and `kernel-64k` for 64k memory pages on aarch64.
```
```suggestion
`""` or `default` for traditional kernel, `realtime` for RT kernel and `64k-pages` for 64k memory pages on aarch64.
```

jbtrystram replied:
good catch !

jlebon reviewed (commented):
I think we also need to have something similar to `queueRevertRTKernel()` but for this new 64k-hugepages kernel.

jbtrystram commented:
I just got this to work nicely on an ARM64 cluster after creating the following machine config : 
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: "worker"
  name: 99-worker-hugepages
spec:
  kernelType: 64k-pages
```

The rendered machineConfig is updated successfully, and the node get updated through the MCD. 
After a reboot I get in a debug pod : 
```
sh-4.4# getconf PAGESIZE
65536
```

jbtrystram commented:
/retest

jbtrystram commented:
Here is the recap of some testing I've done today : 

- Deploy a 4.14 OCP on aarch64
- Upgraded it to a custom release containing this PR + [openshift/os](https://github.com/openshift/os/pull/1351)
- Create the following MachineConfig: 
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: "worker"
  name: 99-worker-hugepages
spec:
  kernelType: 64k-pages
```

A new `MachineConfig` is rendered : `rendered-worker-bf7b27b3afb07f5c8659e0ca73363294`

MCD log :
```
I0914 12:40:31.274415 2442 update.go:1975] Initiating switch to kernel 64k-pages
I0914 12:40:31.276810 2442 update.go:1960] Running: rpm-ostree override remove kernel kernel-core kernel-modules kernel-modules-core kernel-modules-extra --install kernel-64k-core --install kernel-64k-modules --install kernel-64k-modules-core --install kernel-64k-modules-extra
Checking out tree 03d5f14...done
Enabled rpm-md repositories: coreos-extensions
Importing rpm-md...done
rpm-md repo 'coreos-extensions' (cached); generated: 2023-09-07T09:25:22Z solvables: 52
Resolving dependencies...done
Importing packages...done
Applying 5 overrides and 4 overlays
Processing packages...done
Running pre scripts...done
Running post scripts...done
Running posttrans scripts...done
Writing rpmdb...done
Generating initramfs...done
Writing OSTree commit...done
Staging deployment...done
Removed:
kernel-5.14.0-284.30.1.el9_2.aarch64
kernel-core-5.14.0-284.30.1.el9_2.aarch64
kernel-modules-5.14.0-284.30.1.el9_2.aarch64
kernel-modules-core-5.14.0-284.30.1.el9_2.aarch64
kernel-modules-extra-5.14.0-284.30.1.el9_2.aarch64
Added:
kernel-64k-core-5.14.0-284.30.1.el9_2.aarch64
kernel-64k-modules-5.14.0-284.30.1.el9_2.aarch64
kernel-64k-modules-core-5.14.0-284.30.1.el9_2.aarch64
kernel-64k-modules-extra-5.14.0-284.30.1.el9_2.aarch64
Use "rpm-ostree override reset" to undo overrides
Run "systemctl reboot" to start a reboot
I0914 12:43:45.102765 2442 update.go:1975] Rebooting node
I0914 12:43:45.104782 2442 update.go:2005] Removing SIGTERM protection
I0914 12:43:45.104812 2442 update.go:1975] initiating reboot: Node will reboot into config rendered-worker-bf7b27b3afb07f5c8659e0ca73363294
I0914 12:43:45.169801 2442 update.go:1975] reboot successful
I0914 12:43:45.171769 2442 daemon.go:652] Node ip-10-0-9-8.us-east-2.compute.internal is queued for a reboot, skipping sync.
```
[...]
```
I0914 12:44:49.339598 2479 rpm-ostree.go:307] Running captured: rpm-ostree status
I0914 12:44:49.374411 2479 daemon.go:1473] State: idle
Deployments:
* ostree-unverified-registry:quay.io/jbtrystramtestimages/rhcos-container:latest
  Digest: sha256:0d68cf52beba8ef17fd549d22cdff97abe41701549fb778300f0bfb69484e942
  Version: 414.92.202309070913-0 (2023-09-14T12:18:33Z)
  RemovedBasePackages: kernel-modules-core kernel-modules-extra kernel-core kernel kernel-modules 5.14.0-284.30.1.el9_2
  LayeredPackages: kernel-64k-core kernel-64k-modules kernel-64k-modules-core
  kernel-64k-modules-extra
  ostree-unverified-registry:quay.io/jbtrystramtestimages/rhcos-container:latest
  Digest: sha256:0d68cf52beba8ef17fd549d22cdff97abe41701549fb778300f0bfb69484e942
  Version: 414.92.202309070913-0 (2023-09-14T12:18:33Z)
```

- Debug terminal in a node :
```
sh-4.4# getconf PAGESIZE
65536
```

All the nodes are up with the new MachineConfig: 
```
$ oc get mcp                                                                                                                                
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-b355df462a7693e9bc4627068958c679   True      False      False      3              3                   3                     0                      119m
worker   rendered-worker-bf7b27b3afb07f5c8659e0ca73363294   True      False      False      3              3                   3                     0                      119m
```

- revert MachineConfig to default : `oc delete machineconfig 99-worker-hugepages`

MCD log  
```
I0914 13:03:32.024426 2484 update.go:1960] Running: rpm-ostree override reset kernel kernel-core kernel-modules kernel-modules-core kernel-modules-extra --uninstall kernel-64k-core --uninstall kernel-64k-modules --uninstall kernel-64k-modules-core --uninstall kernel-64k-modules-extra
Staging deployment...done
Removed:
kernel-64k-core-5.14.0-284.30.1.el9_2.aarch64
kernel-64k-modules-5.14.0-284.30.1.el9_2.aarch64
kernel-64k-modules-core-5.14.0-284.30.1.el9_2.aarch64
kernel-64k-modules-extra-5.14.0-284.30.1.el9_2.aarch64
Added:
kernel-5.14.0-284.30.1.el9_2.aarch64
kernel-core-5.14.0-284.30.1.el9_2.aarch64
kernel-modules-5.14.0-284.30.1.el9_2.aarch64
kernel-modules-core-5.14.0-284.30.1.el9_2.aarch64
kernel-modules-extra-5.14.0-284.30.1.el9_2.aarch64
Run "systemctl reboot" to start a reboot
```

- Debug terminal in a node :
```
sh-4.4# getconf PAGESIZE
4096
sh-5.1# rpm-ostree status
State: idle
Deployments:
* ostree-unverified-registry:quay.io/jbtrystramtestimages/rhcos-container:latest
  Digest: sha256:0d68cf52beba8ef17fd549d22cdff97abe41701549fb778300f0bfb69484e942
  Version: 414.92.202309070913-0 (2023-09-14T12:18:33Z)
```

All the nodes are back to the previously rendered MachineConfig: 
```
[machine-config-operator] oc get mcp                                              
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-b355df462a7693e9bc4627068958c679   True      False      False      3              3                   3                     0                      151m
worker   rendered-worker-02235dd022ac27fd6988786d72b22f16   True      False      False      3              3                   3                     0                      151m
```

travier commented:
/retest

jbtrystram commented:
/retest

sinnykumari commented on install/0000_80_machine-config-operator_01_machineconfig.crd.yaml:
```diff
@@ -340,7 +340,7 @@ spec:
               kernelType:
                 description: Contains which kernel we want to be running like default
-                  (traditional), realtime
+                  (traditional), realtime, 64k pages
```
I am confused, subject says kernel-64K extension but we are making it a kernelType. Looking at rest of the code it seems to be a new kernelType, so please update subject of the PR accordingly.

sinnykumari replied:
(traditional), realtime, 64k-pages
```suggestion
                  (traditional), realtime, 64k pages
```

sinnykumari replied:
```suggestion
                  (traditional), realtime, 64k-pages
```

sinnykumari reviewed (commented):
Please link jira story associated to this PR.
This will need e2e testing added as well, existing kernelType related test is at https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L195 .
One question: Are we going to have 64K kernelType for RT kernel as well in future or it is irrelevant?
Putting hold for QE pre-merge testing
/hold

sinnykumari commented:
Looking at epic https://issues.redhat.com/browse/COS-2402 and PR description, it seems like this is only relvent to aarch64.  I am curious, are we allowing users to switch to kernel-64k on all arches or only on aarch64? If so, mentioning this explicitly in doc would avoid accidental use of this field on other arches. Also perhaps throwing an error message in daemon log on non aarch64 nodes.

sinnykumari commented:
> This will need e2e testing added as well, existing kernelType related test is at https://github.com/openshift/machine-config-operator/blob/master/test/e2e/mcd_test.go#L195 .

One note, this will make sense when this is supported on x86_64 as well because our current e2e-gcp-op runs on x86_64 nodes. We may need different e2e test added if this is aarch64 exclusive.

travier commented:
As far as I know, this is aarch64 specific and unlikely to come to x86_64 any time soon.

Prashanth684 commented:
> > > One note, this will make sense when this is supported on x86_64 as well because our current e2e-gcp-op runs on x86_64 nodes. We may need different e2e test added if this is aarch64 exclusive.
> > 
> > 
> > Can you give me some guidance for that ?
> 
> Umm, e2e tests config that runs on MCO PRs are in reelase repo, e.g[ https://github.com/openshift/release/ repo. ](https://github.com/openshift/release/blob/master/ci-operator/config/openshift/machine-config-operator/openshift-machine-config-operator-master.yaml) . I couldn't find an example in this repo for e2e tests that would be running on aarch64. Perhaps multi-arch team can help and they can tell better weigh in if having e2e test (on PRs/payload tests) for this new functionality makes sense.
> 
> /cc @Prashanth684

on PR tests are not yet supported for non x86 architectures. There is work on going to get builder images available upstream after which there would be an effort to just pick and choose certain repos where on PR tests would be useful for multi-arch. Meanwhile is there possibility of adding such a job to the nightly runs? cc @jeffdyoung @aleskandro for awareness.

sinnykumari commented:
>  Meanwhile is there possibility of adding such a job to the nightly runs?

Yes, it is possible to add in nightly run. I am not sure where are the configs for aarch64 nightly runs. To test this, all you need to do is on a aarch64 cluster, create a MachineConfig something like https://github.com/openshift/machine-config-operator/pull/3903#issuecomment-1719459537 and apply this to the cluster for all the MCPs you want. Once applied successfully, you would also want to test rollback to default kernel and possible upgrade.

jbtrystram commented on docs/MachineConfiguration.md:
```diff
@@ -201,7 +201,13 @@ spec:
+| 4.6           |  `usbguard`                  |
+| 4.8           |  `usbguard`, `sandboxed-containers`    |
+| 4.11          |  `usbguard`, `sandboxed-containers`, `kerberos`   |
+| 4.14          |  `usbguard`, `sandboxed-containers`, `kerberos`, `ipsec`, `wasm`   |
```
I did not changed the content here, simply reformatted in a table as it was confusing.

jbtrystram replied:
> On the second thought, I would say we need to update master/docs/MachineConfiguration.md#kerneltype as this is kernelType addition.

This is already updated. I tried to add another note to clarify further.

sinnykumari replied:
ah I see. You have made it nicer format. Thanks for doing that :)

jbtrystram commented:
/retest

Prashanth684 commented:
> @sinnykumari I pushed a couple of commits with doc improvements and an error log if `64k-pages` are selected for non aarch64 nodes.
> 
> > > Meanwhile is there possibility of adding such a job to the nightly runs?
> > 
> > 
> > Yes, it is possible to add in nightly run. I am not sure where are the configs for aarch64 nightly runs. To test this, all you need to do is on a aarch64 cluster, create a MachineConfig something like [#3903 (comment)](https://github.com/openshift/machine-config-operator/pull/3903#issuecomment-1719459537) and apply this to the cluster for all the MCPs you want. Once applied successfully, you would also want to test rollback to default kernel and possibly upgrade.
> 
> Is it my responsibility to add that ? @Prashanth684 can you give me some pointers ?

This is where the nightly job configs reside: https://github.com/openshift/release/tree/master/ci-operator/config/openshift/multiarch. The job definitions are here: https://github.com/openshift/release/tree/master/ci-operator/jobs/openshift/multiarch. This probably involves a generic step in the step registry which can then be applied to the respective jobs.

Prashanth684 commented:
> > @sinnykumari I pushed a couple of commits with doc improvements and an error log if `64k-pages` are selected for non aarch64 nodes.
> > > > Meanwhile is there possibility of adding such a job to the nightly runs?
> > > 
> > > 
> > > Yes, it is possible to add in nightly run. I am not sure where are the configs for aarch64 nightly runs. To test this, all you need to do is on a aarch64 cluster, create a MachineConfig something like [#3903 (comment)](https://github.com/openshift/machine-config-operator/pull/3903#issuecomment-1719459537) and apply this to the cluster for all the MCPs you want. Once applied successfully, you would also want to test rollback to default kernel and possibly upgrade.
> > 
> > 
> > Is it my responsibility to add that ? @Prashanth684 can you give me some pointers ?
> 
> This is where the nightly job configs reside: https://github.com/openshift/release/tree/master/ci-operator/config/openshift/multiarch. The job definitions are here: https://github.com/openshift/release/tree/master/ci-operator/jobs/openshift/multiarch. This probably involves a generic step in the step registry which can then be applied to the respective jobs.

Using the arch conditional, I think we can do the same as for rt-kernel here: https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/conf/gcp/ipi-conf-gcp-commands.sh

aleskandro commented:
> > > > One note, this will make sense when this is supported on x86_64 as well because our current e2e-gcp-op runs on x86_64 nodes. We may need different e2e test added if this is aarch64 exclusive.
> > > 
> > > 
> > > Can you give me some guidance for that ?
2> > 
> > 
> > Umm, e2e tests config that runs on MCO PRs are in reelase repo, e.g[ https://github.com/openshift/release/ repo. ](https://github.com/openshift/release/blob/master/ci-operator/config/openshift/machine-config-operator/openshift-machine-config-operator-master.yaml) . I couldn't find an example in this repo for e2e tests that would be running on aarch64. Perhaps multi-arch team can help and they can tell better weigh in if having e2e test (on PRs/payload tests) for this new functionality makes sense.
> > /cc @Prashanth684
> 
> on PR tests are not yet supported for non x86 architectures. There is work on going to get builder images available upstream after which there would be an effort to just pick and choose certain repos where on PR tests would be useful for multi-arch. Meanwhile is there possibility of adding such a job to the nightly runs? cc @jeffdyoung @aleskandro for awareness.

I just spoke to @rioliu-rh... The MCO QE team will run some pre-merge testing with our support and automate some cases in the QE test suites... 

As for the nightly ci tests under the multiarch org, we'd need some new test cases in origin I think, right?

cgwalters commented:
> As for the nightly ci tests under the multiarch org, we'd need some new test cases in origin I think, right?

Yes, this is probably another informing job.

sinnykumari reviewed (commented):
Request one minor change, otherwise lgtm.

sinnykumari commented:
> I am not sure what tests we'd need specific to kernel-64k.

Some of them can be:
1.  switching to 64k-pages kernel and reverting back to 4k kernel.
2. Upgrade of  64k-pages kernel based cluster
Not sure if aarch64 team may have some more fine grained tests in mind too.

sinnykumari reviewed (commented):
/lgtm
Feel free to remove hold once QE has tested and added qe-approved label

cdoern commented:
/hold
Now that: https://github.com/openshift/api/pull/1453 has merged I think we should be modifying the CRD there rather than here.

jbtrystram commented:
> /hold Now that: [openshift/api#1453](https://github.com/openshift/api/pull/1453) has merged I think we should be modifying the CRD there rather than here.

@cdoern see https://github.com/openshift/api/pull/1612

jbtrystram commented:
> > /hold Now that: [openshift/api#1453](https://github.com/openshift/api/pull/1453) has merged I think we should be modifying the CRD there rather than here.
> 
> @cdoern see [openshift/api#1612](https://github.com/openshift/api/pull/1612)

this is now merged :)

jbtrystram commented:
/retest

aleskandro commented:
Hi, this feature cannot be supported on the GCP platforms as their VMs do not allow booting kernels with 64k pagesize.

We'll document this, bit we wonder if we could error out something in case a user tries to apply an MC for that in GCP. I'd see it as a nice to have for improving the UX behind this feature.

Cc @andymcc @Prashanth684 @jeffdyoung

Prashanth684 commented on docs/MachineConfiguration.md:
```diff
@@ -180,7 +180,8 @@ When a machine boots with `nosmt` Kernel Argument, it disables multi-threading o
 This feature is available with OCP 4.4 and onward releases as both `day 1` and `day 2` operation. It allows to choose between traditional and Real Time (RT) kernel on an RHCOS node. Supported values are
-`""` or `default` for traditional kernel and `realtime` for RT kernel.
+`""` or `default` for traditional kernel, `realtime` for RT kernel and `64k-pages` for 64k memory pages on aarch64.
+Note that `64k-pages` and `realtime` cannot be selected at the same time. Also, 64k pages support is limited to aarch64 architecture.
```
is there some way we can limit this change to apply only on aarch64 nodes? and possibly throw some sort of error if the associated machine pool is not aarch64 based?

rioliu-rh replied:
As I know, RT kernel is not supported on ARM64 cluster, So apply kernelType `realtime` + `64k-pages` on ARM64 cluster is invalid case.

If we apply kernelType `64k-pages` on AMD64 cluster, machine config pool will be degraded with below error message 
```console
- lastTransitionTime: '2023-10-18T06:55:53Z'
  message: 'Failed to render configuration for pool worker: kernelType=64k-pages is
    invalid'
  reason: ''
  status: 'True'
  type: RenderDegraded
```

jbtrystram replied:
@Prashanth684  the MCD will throw an error. See https://github.com/openshift/machine-config-operator/pull/3903/files#diff-349c0748c3f52201852d3027c29daf618b45300b949482728df6666a3f9ba245R1145-R1149

Prashanth684 replied:
oops..did not notice that ...thanks

Prashanth684 replied:
also does this error mean that the machine will be unusable when the update fails? i am thinking of a case where we have a multi-arch compute cluster with x86+arm64 compute nodes? would the x86 machines error out and be in a "NotReady" state?

jbtrystram replied:
@Prashanth684 I'll try it out today and followup

aleskandro replied:
> Note that `64k-pages` and `realtime` cannot be selected at the same time. Also, 64k pages support is limited to aarch64 architecture.

Should we also say that `realtime` cannot be selected on `aarch64` to make things clearer (not sure for P/Z)?

rioliu-rh replied:
> also does this error mean that the machine will be unusable when the update fails? i am thinking of a case where we have a multi-arch compute cluster with x86+arm64 compute nodes? would the x86 machines error out and be in a "NotReady" state?

If the `realtime` kernel is applying on arm64 node, the machine will be degraded i.e. `machineconfiguration.openshift.io/state: Degraded` annotation `machineconfiguration.openshift.io/reason` will show you that rt packages are not available

jbtrystram replied:
Clusterbot is not working for me today so I can't try that quickly without going through the process of building a whole release payload, which I won't be able to do today.

sinnykumari replied:
If user applies unsupported combination node will goes degraded. MCO applies config per pool. So, it is admin responsibility to make sure that they apply generic config when a pool has multi architecture nodes. Good thing is if one node fails to apply the config, MCD marks the node degraded and hence it won't start upgrading another node (when maxUnavialble: 1) to cascade the issue. On the degraded node, existing stuff would work fine but in order to schedule a new pod, issue on the node will need to be first resolved so that update can complete and node is marked again schedulable.

rioliu-rh commented:
Update latest test result
executed 4 cases (details can be found in [MCO-798](https://issues.redhat.com/browse/MCO-798) on AWS cluster, all pass
cases are failed on GCP cluster, like mentioned in above comment, CPU does not support `64k-pages` kernel, node cannot be started
```console
EFI stub: ERROR: This 64 KB granular kernel is not supported by your CPU

  Failed to boot both default and fallback entries.

Press any key to continue...
```
MCO QE will execute cases on other platforms later

rioliu-rh commented:
Update latest test result
Executed 4 cases on Azure, no issue found

jlebon commented:
> Hi, this feature cannot be supported on the GCP platform as their VMs do not allow booting kernels with 64k pagesize.

Interesting. Do you have a link about this with more information? Is it for all machine types or only the smaller ones?

> We'll document this, but we wonder if we could error out something in case a user tries to apply an MC for that in GCP. I'd see it as a nice to have for improving the UX behind this feature.

If indeed it's not supported at all, my vote would also be to give a nicer error here rather than wait until the user hits a likely much more obscure error down the line. That said, I wouldn't necessarily block this PR on this if @jbtrystram would rather do that as a follow-up.

aleskandro commented:
> > Hi, this feature cannot be supported on the GCP platform as their VMs do not allow booting kernels with 64k pagesize.
> 
> Interesting. Do you have a link about this with more information? Is it for all machine types or only the smaller ones?
> 

Any machine size cannot boot a 64k pagesize Linux kernel on Google Cloud. We have some internal discussions about this, and there should BZs. I'll try to dig them again.

rioliu-rh commented:
Update latest test result
@sergiordlr did test on both baremetal HA/SNO cluster, no issue found.

Summarize the test status
- tests are executed on AWS, Azure, GCP and Baremetal platforms
- tests are passed on AWS, Azure, Baremetal platforms
- tests are failed on GCP platform, CPU used by GCP VM does not support 64k page size kernel (Doc needed)

rioliu-rh commented:
/unhold

rioliu-rh commented:
/label qe-approve

rioliu-rh commented:
/label qe-approved

sinnykumari commented:
QE testing has been done.
/lgtm
